Fix all uninitialized variable warnings (C26495)

Pokechu22 opened the pull request:
This fixes all occurrences of [C26495](https://docs.microsoft.com/en-us/cpp/code-quality/c26495?view=msvc-160) within Dolphin code, by adding initializers for everything.  I have mixed feelings about this approach, but it seems like there isn't any simpler approach (a default constructor won't work for instance).

See https://github.com/dolphin-emu/dolphin/pull/9956#issuecomment-911211718 for an example of where a value not being initialized caused problems.

sepalani commented on Source/Core/Core/BootManager.cpp:
```diff
@@ -64,42 +64,42 @@ struct ConfigCache
+  bool bSyncGPU = false;
+  int iSyncGpuMaxDistance = false;
+  int iSyncGpuMinDistance = false;
+  float fSyncGpuOverclock = false;
```
Shouldn't this be initialised with `0.0f` instead of false?

Pokechu22 replied:
Yes, these are copy-paste errors.  I think initializing a float with `0` instead of `0.0f` is fine, but `false` is definitely silly.

sepalani reviewed (commented):
Some remarks:
 - Is there a reason to initialise some integers/bool/float types with `{}` where as in other places it uses an assignation `=`?
 - Can't braces be used everywhere to keep it consistent?
 - Should float be initialised with `0.0f` rather than `0` or `false`?

BhaaLseN commented on Source/Core/Core/DSP/DSPCore.h:
```diff
@@ -279,10 +279,10 @@ struct DSP_Regs
   // DSP IROM blob, which is where the DSP boots from. Embedded into the DSP.
-  std::array<u16, DSP_IROM_SIZE> irom_contents;
+  std::array<u16, DSP_IROM_SIZE> irom_contents = {};
 
   // DSP DROM blob, which contains resampling coefficients.
-  std::array<u16, DSP_COEF_SIZE> coef_contents;
+  std::array<u16, DSP_COEF_SIZE> coef_contents = {};
```
Those two seem to be outliers, other code uses `std::array{}` while this uses `std::array = {}`. I think we should be consistent, if the compiler lets us.

BhaaLseN commented on Source/Core/Core/DSP/Jit/x64/DSPJitRegCache.h:
```diff
@@ -170,8 +170,8 @@ class DSPJitRegCache
 
-  std::array<DynamicReg, 37> m_regs;
-  std::array<X64CachedReg, 16> m_xregs;
+  std::array<DynamicReg, 37> m_regs = {};
+  std::array<X64CachedReg, 16> m_xregs = {};
```
Same down here.

BhaaLseN commented on Source/Core/Core/DSP/Jit/x64/DSPJitRegCache.cpp:
```diff
@@ -194,7 +194,7 @@ DSPJitRegCache::DSPJitRegCache(DSPEmitter& emitter)
 DSPJitRegCache::DSPJitRegCache(const DSPJitRegCache& cache)
     : m_regs(cache.m_regs), m_xregs(cache.m_xregs), m_emitter(cache.m_emitter),
-      m_is_temporary(true), m_is_merged(false)
+      m_is_temporary(true), m_is_merged(false), m_use_ctr(0)
```
Thoughts on moving everything that uses a constant value into the header (with direct initialization)?

BhaaLseN commented on Source/Core/Core/Debugger/Dump.cpp:
```diff
@@ -9,7 +9,7 @@
 #include "Common/IOFile.h"
 
-CDump::CDump(const std::string& filename) : m_pData(nullptr)
+CDump::CDump(const std::string& filename) : m_size(0), m_pData(nullptr)
```
Same here for members that are initialized to constant values

BhaaLseN commented on Source/Core/Core/FifoPlayer/FifoDataFile.h:
```diff
@@ -93,13 +94,13 @@ class FifoDataFile
-  u8 m_TexMem[TEX_MEM_SIZE];
-  u32 m_ram_size_real;
-  u32 m_exram_size_real;
+  std::array<u32, BP_MEM_SIZE> m_BPMem;
+  std::array<u32, CP_MEM_SIZE> m_CPMem;
+  std::array<u32, XF_MEM_SIZE> m_XFMem;
+  std::array<u32, XF_REGS_SIZE> m_XFRegs;
+  std::array<u8, TEX_MEM_SIZE> m_TexMem;
```
Do we also want those to be initialized?

BhaaLseN commented on Source/Core/Core/HW/WiiSaveStructs.h:
```diff
@@ -91,8 +91,8 @@ class Storage
     };
-    u8 mode, attributes;
-    Type type;
+    u8 mode = 0, attributes = 0;
```
Would you mind splitting those two into two lines?

BhaaLseN commented on Source/Core/Core/PowerPC/MMU.h:
```diff
@@ -49,7 +49,15 @@ struct TryReadResult
 
+#ifdef _MSC_VER
+#pragma warning(push)
+// Disable warning for uninitialized member variables
```
Can you add a reason _why_ this warning is temporarily disabled for this ctor? (Or: simply address it, so the suppression isn't needed)

Pokechu22 replied:
It's because this class is kinda like `std::optional`.  Probably a better solution would be to replace it with `std::optional` to guarantee nobody can touch the other fields.

Pokechu22 replied:
I've replaced `TryReadResult` and `TryWriteResult` with `std::optional<ReadResult>` and `std::optional<WriteResult>`.

BhaaLseN commented on Source/Core/Core/PowerPC/PPCAnalyst.h:
```diff
@@ -138,23 +138,23 @@ using CodeBuffer = std::vector<CodeOp>;
 
   // Register statistics about the block.
-  BlockRegStats *m_gpa, *m_fpa;
+  BlockRegStats *m_gpa = nullptr, *m_fpa = nullptr;
```
Can we break those up into two lines?

BhaaLseN commented on Source/Core/VideoCommon/GXPipelineTypes.h:
```diff
@@ -35,10 +35,18 @@ struct GXPipelineUid
   GXPipelineUid() { std::memset(static_cast<void*>(this), 0, sizeof(*this)); }
+#ifdef _MSC_VER
+#pragma warning(push)
+// Disable warning for uninitialized member variables
```
Can you add a reason to this comment, or address it so it isn't necessary?
My guess is that VS doesn't recognize the `memcpy` as initialization?

Pokechu22 replied:
Yep, it's because it doesn't recognize memcpy (but it _does_ recognize memset?)

BhaaLseN commented on Source/Core/VideoCommon/TextureCacheBase.h:
```diff
@@ -89,28 +89,28 @@ class TextureCacheBase
-        native_height;  // Texture dimensions from the GameCube's point of view
-    unsigned int native_levels;
+    // Texture dimensions from the GameCube's point of view
+    unsigned int native_width = 0, native_height = 0;
```
Can we break those up into two lines?

BhaaLseN reviewed (commented):
Changes seem mostly fine, but I'm a bit worried about consistency. Some places use assignments, others use uniform initialization for the same thing; and mixing those two all over the place may lead to more inconsistencies down the road (depending on what file one looks at for inspiration).
Then again, this is likely something we should first address in the code style docs before we force one over the other on a whim.

I do wonder about those `int i = false`, `float f = false` etc. though that sepalani pointed out, seems like copy&paste gone wrong.

sepalani commented on Source/Core/Core/DSP/Jit/x64/DSPJitRegCache.h:
```diff
@@ -170,14 +170,14 @@ class DSPJitRegCache
   DSPEmitter& m_emitter;
-  bool m_is_temporary;
-  bool m_is_merged;
+  bool m_is_temporary = true;
```
Depending on the constructor `m_is_temporary` is set to `true` or `false`. Won't setting it here produce side effect?

Pokechu22 replied:
Per [cppreference](https://en.cppreference.com/w/cpp/language/constructor), starting in C++11,

> If a non-static data member has a default member initializer and also appears in a member initializer list, then the member initializer is used and the default member initializer is ignored

So, it's fine, but maybe unnecessary.

sepalani commented on Source/Core/DolphinQt/Config/Mapping/IOWindow.cpp:
```diff
@@ -207,7 +207,7 @@ class InputStateLineEdit : public QLineEdit
 private:
   std::function<ControlState()> m_state_evaluator;
-  bool m_should_paint_state_indicator;
+  bool m_should_paint_state_indicator = true;
```
Shouldn't it be initialised to false?

Pokechu22 replied:
I'm not sure.  I based it on this code:

https://github.com/dolphin-emu/dolphin/blob/4b8b53ac732645f6c0fd3f4b11733cb3d86c9c41/Source/Core/DolphinQt/Config/Mapping/IOWindow.cpp#L670-L689

which sets it to `true` in the default case.  But maybe `false` would be better since the default should be empty.

sepalani commented on Source/Core/Common/IOFile.h:
```diff
@@ -57,6 +58,18 @@ class IOFile
+  }
+
+  template <typename T, std::size_t N>
+  bool WriteArray(std::array<T, N>& elements)
```
Can a const reference be used?

iwubcode commented on Source/Core/VideoCommon/VideoConfig.h:
```diff
@@ -58,173 +58,173 @@ struct VideoConfig final
-  AspectMode suggested_aspect_mode;
-  bool bCrop;  // Aspect ratio controls.
-  bool bShaderCache;
+  bool bVSync{};
```
Throughout most of the files you used `bool x = false` or `int x = 0`.  Here you use `bool x{}` and `int x{}`.  Can we make this consistent?

Pokechu22 replied:
I've changed them to `= false` or `= 0`.

iwubcode commented on Source/Core/VideoCommon/ShaderCache.h:
```diff
@@ -211,16 +211,16 @@ class ShaderCache final
-  ShaderModuleCache<PixelShaderUid> m_ps_cache;
-  ShaderModuleCache<UberShader::VertexShaderUid> m_uber_vs_cache;
-  ShaderModuleCache<UberShader::PixelShaderUid> m_uber_ps_cache;
+  ShaderModuleCache<VertexShaderUid> m_vs_cache{};
```
If you move `m_api_type` value to the header.  I think you can then remove or default `ShaderCache()`.  Then the `{}` for all of these 'complex types' can probably go away?

Pokechu22 replied:
They aren't needed even without it; I'll remove it.  I think I added them before the `bool pending = false;` in `ShaderModuleCache::Shader`.  (Note that it takes quite some time for the analysis to run, and the analysis doesn't always include all of the warnings if there's more than one for a given file, so I may be initializing things that didn't need to be explicitly initialized).

iwubcode commented on Source/Core/AudioCommon/OpenALStream.h:
```diff
@@ -68,9 +68,9 @@ class OpenALStream final : public SoundStream
-  ALfloat m_volume;
+  std::array<ALuint, OAL_BUFFERS> m_buffers{};
+  ALuint m_source = 0;
+  ALfloat m_volume = 1;
```
I know nothing about this system.  Is 1 intentional here?

Pokechu22 replied:
It's used here and here:

https://github.com/dolphin-emu/dolphin/blob/05f86b9f5b6e7f65aaefcc29e348cc2628bed34f/Source/Core/AudioCommon/OpenALStream.cpp#L259-L260

https://github.com/dolphin-emu/dolphin/blob/05f86b9f5b6e7f65aaefcc29e348cc2628bed34f/Source/Core/AudioCommon/OpenALStream.cpp#L147-L153

And the docs say 1.0 leaves volume unchanged:

https://github.com/dolphin-emu/dolphin/blob/05f86b9f5b6e7f65aaefcc29e348cc2628bed34f/Externals/OpenAL/include/al.h#L184-L196

iwubcode commented on Source/Core/Core/PowerPC/MMU.h:
```diff
@@ -37,49 +37,38 @@ u32 HostRead_Instruction(u32 address);
-  {
-  }
-  explicit operator bool() const { return success; }
+  ReadResult(bool translated_, T&& value_) : translated(translated_), value(std::move(value_)) {}
```
I may be wrong here.  But I think since this is a `T&&` this line should be:

`ReadResult(bool translated_, T&& value_) : translated(translated_), value(std::forward<T>(value_)) {}`

The reason is T&& can be either an lvalue or an rvalue.  But move is only applicable on rvalues.

(I know you didn't write this code originally but just thought I'd mention it)

Pokechu22 replied:
Changed.

iwubcode commented on Source/Core/Core/PowerPC/MMU.h:
```diff
@@ -93,36 +82,37 @@ void HostWrite_U64(u64 var, u32 address);
-  TryWriteResult() : success(false) {}
-  TryWriteResult(bool translated_) : success(true), translated(translated_) {}
-  explicit operator bool() const { return success; }
+  WriteResult(bool translated_) : translated(translated_) {}
```
Surprised this wasn't `explicit`

Pokechu22 replied:
Changed.

sepalani reviewed (commented):
> Reviewed 1 of 2 files at r1, 110 of 110 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/10085)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @Pokechu22)



<!-- Sent from Reviewable.io -->

sepalani reviewed (commented):
I don't think making this PR stale for too long is a good idea because it will be a pain to maintain.

PR tested on Windows debug build with some games. LGTM besides these 2 nitpicks.

delroth commented:
@dolphin-emu-bot rebuild

Pokechu22 commented on Source/Core/Common/IOFile.h:
```diff
@@ -57,6 +58,18 @@ class IOFile
   }
 
+  template <typename T, std::size_t N>
+  bool ReadArray(std::array<T, N>& elements, size_t* num_read = nullptr)
```
I was mostly modeling it around `ChunkFile` (the magic code used for save-states), but probably it's better to not do that.  `ChunkFile` _needs_ to be magic, but IOFile doesn't.

Pokechu22 commented:
Rebased and updated.  I also had to replace `TryReadResult` with `std::optional` in CheatSearch since #9886 was merged; that commit should be re-reviewed.

sepalani reviewed (approved):
LGTM

BhaaLseN commented on Source/Core/VideoCommon/VideoConfig.h:
```diff
@@ -58,173 +58,174 @@ struct VideoConfig final
+  bool bSkipPresentingDuplicateXFBs = false;
+  bool bCopyEFBScaled = false;
+  int iSafeTextureCache_ColorSamples = 0;
   float fAspectRatioHackW, fAspectRatioHackH;
```
Did we miss those two, or are they set elsewhere?

Plus, it would probably be a good idea to make this into two lines while we're at it.

Pokechu22 replied:
They're set in the constructor.  But so are a few other things which I did initialize - I'll move those values out of the constructor.

Pokechu22 commented:
Note: The fifoci comment is because for some reason it compared with 5.0-15257 even though this PR was based on 5.0-15259 (#10138).  The [more recent comparision](https://fifo.ci/version/23e81202162c52ee9854feb9b72c3aa0424fc793/) has no differences (though the same things still fail to render, which is expected), but fifoci doesn't delete old comments if there are no changes.

shuffle2 commented:
@Pokechu22  I just looked for uninit vars the same way again. Among a few we don't really care about (but probably should just initialize and let the compiler deal with dead store elimination), there's also: 
```
Severity	Code	Description	Project	File	Line	Suppression State	Detail Description
Warning	C6001	Using uninitialized memory 'scanline'.	DolphinLib	C:\src\dolphin\Source\Core\VideoBackends\Software\EfbInterface.cpp	615	
```
![image](https://user-images.githubusercontent.com/113063/149723674-ad74b0c3-3747-4bf5-9771-21d3517e0ea6.png)


It has correctly detected that if `source_rect.left >= source_rect.right`, then body of `for (int i = 1, x = left; x < right; i++, x++)` won't be executed, and `scanline` won't be initialized. Is that a condition we know never can happen (particularly the equals case)? Should something be done about it?

Pokechu22 commented:
It's impossible for `left >= right` since the copy size is specified as a width/height (and those are unsigned, with a value of 0 indicating 1 pixel).  See this code (which also has a rather dubious comment):

https://github.com/dolphin-emu/dolphin/blob/e62771856092cd9bbc25c8c94c433ed0e55a9f3f/Source/Core/VideoCommon/BPStructs.cpp#L225-L228

So, even if a game for some reason decided that the XFB should be as thin as possible, it'd still render correctly (at least here; I don't know what hardware will actually do when you try to put a 1-pixel wide framebuffer to screen, but that'd be a VideoInterface problem I think).

shuffle2 commented:
OK, thanks for pointing out caller.

unrelated, but interesting that their source is 10 bit integers...wonder what happens on 10bit overflow or so (which we wouldn't detect afaict) :p

Pokechu22 commented:
Well, the max texture size is 1024 by 1024, which is the biggest size for a 10-bit value, but the EFB itself is 640 by 528, so a full-sized one would include out of bounds data.  I suspect that an overflow would just wrap around (that's roughly what happens without of bounds scissor values, though that also has wrapping at 2048 for other aspects), but I don't know what actually happens in the 640-1023 region (though apparently that sometimes _does_ happen with Mario Kart Wii).  Dolphin currently clamps it to stay within the 640 by 528 region, I guess leaving out-of-bounds memory untouched, so that overflow wouldn't cause any problems.
